### DISCUSSION TITLE
Add support for custom rulesets to Inso

### DISCUSSION THREAD

**Discussion by mheap** [Kong/insomnia#4114] [Ideas]
Hello! I'd like to extend `inso` with the concept of custom rulesets. It looks as though [linting is performed by Spectral](https://github.com/Kong/insomnia/blob/cbc1cfc8b8e86822ec5303ff1f8310e2795c8258/packages/insomnia-inso/package.json#L71) and with that in mind I'd like to expose Spectral's ability to [set custom rulesets](https://meta.stoplight.io/docs/spectral/ZG9jOjYyMDc0NA-rulesets) through `inso`.

I considered putting the ruleset in the `.insorc` file initially, but this would not allow us to have different rules per API when working with legacy systems. I looked in to how Insomnia stores the API specs as I intended to put the rulesets alongside the specification, but it looks like they're stored in a database.

~~To keep things easy, I'm proposing an `~/.insomnia/rulesets` folder that will contain rulesets. Running `inso lint spec <file> --ruleset my-custom-ruleset` would load `~/.insomnia/rulesets/my-custom-ruleset.yml`. This does not follow the `$XDG_CONFIG_HOME` rules which would have me store the rulesets in `~/Library/Application Support/Insomnia/rulesets`, but is more expected for consumers imo.~~ (This won't work due to syncing)

I'm recommending that we add another database `insomnia.ApiSpecRulesets` for storing custom rulesets. I'll add a new `Rulesets` panel to Design workspaces. ~~Eventually you should be able to configure these in Insomnia, but for now I'm proposing an `inso lint add-ruleset <name> </path/to/file.yml>` to populate the table.~~ These rulesets would also need persisting into the `.insomnia` folder in a repo to be run with CI/CD.

~~Finally, there are custom functions. Similarly to rulesets, I'm proposing an `inso lint add-function <name> </path/to/file.js>` method to load these in to the database before they can be referenced in a ruleset.~~

tl;dr:

- [x] UI to add new rulesets in Insomnia
- [x] Export new rulesets to `.insomnia` export
- [x] Use the new ruleset if it exists in an export. Look up any rulesets in the same workspace ID as the specification being linted
- [x] Make Insomnia use the new ruleset for inline linting
- [ ] Ensure that rulesets can reference named functions
- [x] Add support for exporting rulesets and custom functions into the `.insomnia` folder with `git` sync

**Comment by ikenna** [comment]
Would be good to have this.
